examtool/examtool/api/utils.py

```diff
@@ -28,3 +24,39 @@ def dict_to_list(d):
 
 def list_to_dict(l):
     return {i: x for i, x in enumerate(l)}
+
+
+class IDFactory:
+    def __init__(self, id_start="", id_end="", length=32, allow_random_ids=True):
+        self.length = length
+        self.id_start = id_start
+        self.id_end = id_end
+        self.allow_random_ids = allow_random_ids
+        self.current_ids = set()
+
+    def rand_id(self, length=None):
+        if length is None:
+            length = self.length
+        return "".join(random.choices(string.ascii_uppercase, k=length))
+
+    def id_from_str(self, string):
+        return self.id_start + string + self.id_end
+
+    def get_id(self, string=None):
+        if string is None:
+            if not self.allow_random_ids:
+                raise SyntaxError(
+                    "A custom ID is required but was not set for this question!"
+                )
+            qid = self.rand_id()
+            while qid in self.current_ids:
+                qid = self.rand_id()
+        elif isinstance(string, str):
```

Comment:
no need for the check, if the user calls it badly they can suffer. If you want, put a type annotation

Reply:
Wait why shouldn't we check? It is just to make sure there are no duplicates.

Reply:
I meant no need for the type check, either. Just do `def get_id(self, string: Optional[str] = None)` and make it `if string is None: ... else: ...`

Reply:
Done